internal/statemachine/state_machine.go

```diff
@@ -329,14 +335,18 @@ func (stateMachine *StateMachine) warnUsageOfSystemLabel(volumeName string, stru
 	}
 }
 
-// checkSystemSeed checks if the struture is a system-seed one and fixes the Label if needed
-func (stateMachine *StateMachine) checkSystemSeed(volume *gadget.Volume, structure *gadget.VolumeStructure, structIndex int) {
-	if structure.Role == gadget.SystemSeed {
-		stateMachine.IsSeeded = true
-		if structure.Label == "" {
-			structure.Label = structure.Name
-			volume.Structure[structIndex] = *structure
-		}
+// handleSystemSeed checks if the struture is a system-seed one and fixes the Label if needed
+func (stateMachine *StateMachine) handleSystemSeed(volume *gadget.Volume, structure *gadget.VolumeStructure, structIndex int) {
+	if structure.Role != gadget.SystemSeed {
+		return
+	}
+	stateMachine.IsSeeded = true
+	// The "main" volume is the one with a system-seed structure
+	stateMachine.MainVolumeName = structure.VolumeName
```

Comment:
A follow up to the previous question about usage of `MainVolumeName`, it looks like it's only set for the Ubuntu Core case (where we have seeded images). Is this field only somehow used in the UC20+ case?

Reply:
So the reason why I commented on this and mentioned UC20+ images is that basically we set no `MainVolumeName` for classic cases. For non core images we have no `system-seed` partitions (at least the UC20+ ones, since UC18 also didn't use the seed).
I like the idea to consider the volume with the seed as the main volume, but in this case we basically say that we 'do not consider any volume as main for other cases'. What I was thinking that maybe, before we get better logic, for other cases we should consider the volume that carries the `system-data` partition to be the main one.

Not that it really matters as, like you mentioned in the other comment, this is not currently used. But I'd like us to be consistent and use it throughout, as this is done in a step that is not core-specific.

Reply:
Correct, I forgot about that! I agree that making the volume holding the `system-data` the main one seems to make sense. But currently the goal of identifying the main volume is to know where (in the future) to add the system-data structure if it is missing. So in the classic case, intentionally identifying the main volume by looking for the `system-data` role is not very useful (it is a kind of a deadlock :)). We also already do that and storing it in the `rootfsSeen` var and then use it in `fixMissingSystemData()`. Currently we also check there is only one volume in the gadget info, before adding the new `system-data` structure.

So I suppose the next step would be to find a better heuristic to identify the main volume and be able to add a system-data to the right one.

I agree having a consistent behavior is good, but here I think that would be a bit confusing. What do you think?

Reply:
Okay, we can discuss this further when the need for `MainVolumeName` appears!